src/drizzleReactHooks.js

```diff
@@ -0,0 +1,203 @@
+import React, {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+  useState
+} from 'react'
+import PropTypes from 'prop-types'
+import debounce from 'debounce'
+import shallowequal from 'shallowequal'
+
+const Context = createContext()
+export const useDrizzle = () => useContext(Context)
+export const useDrizzleState = (mapState, args) => {
+  const { drizzle } = useDrizzle()
+  const mapStateRef = useRef(mapState)
+  mapStateRef.current = mapState
+  const argsRef = useRef(args)
+  const [state, setState] = useState(
+    mapStateRef.current(drizzle.store.getState())
+  )
+  const stateRef = useRef(state)
+  if (!shallowequal(argsRef.current, args)) {
+    argsRef.current = args
+    const newState = mapStateRef.current(drizzle.store.getState())
+    if (!shallowequal(stateRef.current, newState)) {
+      stateRef.current = newState
+      setState(newState)
+    }
+  }
+  useEffect(
+    () => {
+      const debouncedHandler = debounce(() => {
+        const newState = mapStateRef.current(drizzle.store.getState())
+        if (!shallowequal(stateRef.current, newState)) {
+          stateRef.current = newState
+          setState(newState)
+        }
+      })
+      const unsubscribe = drizzle.store.subscribe(debouncedHandler)
+      return () => {
+        unsubscribe()
+        debouncedHandler.clear()
+      }
+    },
+    [drizzle.store]
+  )
+  return stateRef.current
+}
+
+export const DrizzleProvider = ({ children, drizzle }) => {
+  const useCacheCall = useCallback(
+    (contractNameOrContractNames, methodNameOrFunction, ...args) => {
+      const isFunction = typeof methodNameOrFunction === 'function'
+      const drizzleState = useDrizzleState(drizzleState => {
+        if (isFunction)
+          return contractNameOrContractNames.reduce((acc, contractName) => {
+            acc[contractName] = drizzleState.contracts[contractName]
+            return acc
+          }, {})
+        else {
+          const cacheKey = drizzle.contracts[
```

Comment:
I'd refactor with more intermediate variables here. Something like:

```js
const instance = drizzle.contracts[contractNameOrContractNames]
const cacheKey = instance.methods[methodNameOrFunction].cacheCall(...args)
```

Consider refactoring with this pattern in other parts of this code.